### PR TITLE
Add safe_strncpy function

### PR DESCRIPTION
**Pull request by janisozaur**
safe_strncpy is similar to strncpy, but makes sure buffer is
null-terminated. Update most of project to use this updated function to
prevent illegal memory accesses.

This PR fixes quite a lot of issues, but not all of them. There are still parts, where strcat could step out of owned ranges, this will come later.

### CODE REVIEW

File: src/windows/options.c [hunk 1]
```diff
@@ -1040,7 +1041,7 @@ static void window_options_dropdown(rct_window *w, int widgetIndex, int dropdown
 					Mixer_Init(devicename);
 					SafeFree(gConfigSound.device);
 					gConfigSound.device = malloc(strlen(devicename) + 1);
-					strcpy(gConfigSound.device, devicename);
+					safe_strncpy(gConfigSound.device, devicename, strlen(devicename) + 1);
```
Comment (IntelOrca) [inline src/windows/options.c]: Probably better to use strdup here or at least our own version.

File: src/util/util.c [hunk 2]
```diff
@@ -181,6 +181,18 @@ int strcicmp(char const *a, char const *b)
 	}
 }
 
+char *safe_strncpy(char * destination, const char * source, size_t num)
+{
+	char *result = strncpy(destination, source, num);
+	int len = strnlen(source, num);
```
Comment (IntelOrca) [inline src/util/util.c]: To do both strncpy and strnlen is inefficient - I don't the compiler can optimise this out either. So might be better to integrate our own loop.

File: src/windows/options.c (hunk 1 above)
Comment (janisozaur) [inline src/windows/options.c]: Well, it pretty much is strdup now…

Comment (IntelOrca) [inline src/windows/options.c]: You need to change it to use strdup.

Comment (janisozaur) [inline src/windows/options.c]: done

File: src/util/util.c (hunk 2 above)
Comment (janisozaur) [inline src/util/util.c]: done

**Comment by janisozaur** [comment]
ok, so VS has no `strndup` at all. Can you tell me how you want it fixed then? Can we just go to `safe_strncpy`?

**Comment by IntelOrca** [comment]
@janisozaur make a function
```
char *strcpyalloc(const char *src, size_t maxSrcLength)
{
    int len = strnlen(src, maxSrcLength);
    char *result = malloc(len + 1);
    memcpy(result, src, len);
    result[len] = 0;
    return result;
}
```

**Comment by janisozaur** [comment]
is it ok now?

**Comment by IntelOrca** [comment]
yes

**Pull request merged**